### PR TITLE
KAFKA-9568: enforce rebalance if client endpoint has changed

### PR DESCRIPTION
**Pull request by ableegoldman**
Since the assignment info includes a map with all member's host info, we can just check the received map to make sure our endpoint is contained. If not, we need to force the group to rebalance and get our updated endpoint info.

### CODE REVIEW

**Comment by ableegoldman** [comment]
call for review @abbccdda @vvcephei

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java [hunk 1]
```diff
@@ -1403,17 +1404,58 @@ public void shouldNotLoopInfinitelyOnMissingMetadataAndShouldNotCreateRelatedTas
 
     @Test
     public void shouldUpdateClusterMetadataAndHostInfoOnAssignment() {
```
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:1286]: This is unrelated to this change, I just happened to notice this test didn't seem to actually be testing anything at all

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignorError.java [hunk 2]
```diff
@@ -19,7 +19,7 @@
 public enum AssignorError {
     NONE(0),
     INCOMPLETE_SOURCE_TOPIC_METADATA(1),
-    VERSION_PROBING(2);
+    REBALANCE_NEEDED(2);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignorError.java:22]: We can reuse this for KIP-441 as well

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 3]
```diff
@@ -496,9 +496,7 @@ private void runLoop() {
         while (isRunning() || taskManager.isRebalanceInProgress()) {
             try {
                 runOnce();
-                if (assignmentErrorCode.get() == AssignorError.VERSION_PROBING.code()) {
-                    log.info("Version probing detected. Rejoining the consumer group to trigger a new rebalance.");
-
+                if (assignmentErrorCode.get() == AssignorError.REBALANCE_NEEDED.code()) {
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:499]: Are we deprecating the VERSION_PROBING?

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:499]: This isn't a client or externally-facing config, just something each thread sets internally if it detects version probing during `onAssignment`
If you're asking about the old log message, I moved it to where we set the error code.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java [hunk 4]
```diff
@@ -1403,17 +1404,58 @@ public void shouldNotLoopInfinitelyOnMissingMetadataAndShouldNotCreateRelatedTas
...
+        EasyMock.replay(taskManager, streamsMetadataState);
+
+        partitionAssignor.onAssignment(createAssignment(initialHostState), null);
+        partitionAssignor.onAssignment(createAssignment(newHostState), null);
+
+        EasyMock.verify(taskManager, streamsMetadataState);
+    }
+
+    @Test
+    public void shouldTriggerRebalanceOnHostInfoChangeOnly() {
```
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java]: Remove `only` maybe?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java [hunk 5]
```diff
@@ -1403,17 +1404,58 @@ public void shouldNotLoopInfinitelyOnMissingMetadataAndShouldNotCreateRelatedTas
...
+        partitionAssignor.onAssignment(createAssignment(oldHostState), null);
+
+        assertThat(partitionAssignor.assignmentErrorCode(), is(AssignorError.REBALANCE_NEEDED.code()));
+
+        partitionAssignor.setAssignmentErrorCode(AssignorError.NONE.code());
+        partitionAssignor.onAssignment(createAssignment(newHostState), null);
+
+        assertThat(partitionAssignor.assignmentErrorCode(), not(AssignorError.REBALANCE_NEEDED.code()));
 
         EasyMock.verify(taskManager);
```
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:1336]: Could we also verify `streamsMetadataState`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 6]
```diff
@@ -110,20 +107,7 @@ public int hashCode() {
...
-                    throw new ConfigException(
-                        String.format("Error parsing host address %s. Expected format host:port.", endPoint)
-                    );
-                }
-
-                hostInfo = new HostInfo(host, port);
-            } else {
-                hostInfo = null;
-            }
+            hostInfo = endPoint != null ? HostInfo.buildFromEndpoint(endPoint) : null;
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: nit: I feel we could internalize the hostInfo creation, including the null check:
`hostInfo = HostInfo.buildFromEndpoint(endPoint);`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 7]
```diff
@@ -1392,6 +1398,10 @@ protected void setAssignmentErrorCode(final Integer errorCode) {
         assignmentErrorCode.set(errorCode);
     }
 
+    Integer assignmentErrorCode(){
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: nit: space after `)`

File: streams/src/main/java/org/apache/kafka/streams/state/HostInfo.java [hunk 8]
```diff
@@ -46,6 +50,22 @@ public HostInfo(final String host,
         this.port = port;
     }
 
+    /**
+     *
+     * @throws ConfigException if the host or port cannot be parsed from the given endpoint string
+     */
+    public static HostInfo buildFromEndpoint(final String endPoint) {
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/state/HostInfo.java:57]: Have you also migrated the unit test?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java [hunk 9]
```diff
@@ -1403,17 +1404,58 @@ public void shouldNotLoopInfinitelyOnMissingMetadataAndShouldNotCreateRelatedTas
...
+        );
+
+        streamsMetadataState = EasyMock.createStrictMock(StreamsMetadataState.class);
         createMockTaskManager();
-        EasyMock.replay(taskManager);
-        streamsMetadataState = EasyMock.createNiceMock(StreamsMetadataState.class);
         configurePartitionAssignor(emptyMap());
 
-        partitionAssignor.onAssignment(createAssignment(hostState), null);
+        streamsMetadataState.onChange(EasyMock.anyObject(), EasyMock.anyObject(), EasyMock.anyObject());
```
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java]: Combine L1421-1422 with `times(2)`?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java [hunk 10]
```diff
@@ -1403,17 +1404,58 @@ public void shouldNotLoopInfinitelyOnMissingMetadataAndShouldNotCreateRelatedTas
...
+            mkEntry(new HostInfo("localhost", 9090), mkSet(t1p0, t1p1)),
+            mkEntry(new HostInfo("otherhost", 9090), mkSet(t2p0, t2p1))
+        );
+
+        final Map<HostInfo, Set<TopicPartition>> newHostState = mkMap(
+            mkEntry(new HostInfo("newhost", 9090), mkSet(t1p0, t1p1)),
+            mkEntry(new HostInfo("otherhost", 9090), mkSet(t2p0, t2p1))
+        );
+
+        streamsMetadataState = EasyMock.createNiceMock(StreamsMetadataState.class);
```
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java]: `streamsMetadataState` never been used?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java [hunk 11]
```diff
@@ -1403,17 +1404,58 @@ public void shouldNotLoopInfinitelyOnMissingMetadataAndShouldNotCreateRelatedTas
...
+        EasyMock.replay(taskManager);
+
+        partitionAssignor.onAssignment(createAssignment(oldHostState), null);
+
+        assertThat(partitionAssignor.assignmentErrorCode(), is(AssignorError.REBALANCE_NEEDED.code()));
+
+        partitionAssignor.setAssignmentErrorCode(AssignorError.NONE.code());
+        partitionAssignor.onAssignment(createAssignment(newHostState), null);
+
+        assertThat(partitionAssignor.assignmentErrorCode(), not(AssignorError.REBALANCE_NEEDED.code()));
```
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java]: Could we just check if the error code is null?

**Review by abbccdda — COMMENTED** [review]
Thanks for the PR. Left a couple of comments.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java (hunk 5 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:1336]: `streamsMetadataState` isn't really part of this test, we just have to create the mock as part of the setup for the assignor configuration

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java (hunk 6 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Ack

File: streams/src/main/java/org/apache/kafka/streams/state/HostInfo.java (hunk 8 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/state/HostInfo.java:57]: I don't think we need to migrate all the unit tests, since we just pass in a host+port directly (vs building it from a single string). But I'll add unit tests for the static builder

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java (hunk 9 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java]: Modified to expect the actual host state

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java (hunk 10 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java]: It's used during the assignor configuration, you have to set it up in all these tests. I just opened a PR to clean this up so once it's merged I'll rebase this PR and remove this here -- see https://github.com/apache/kafka/pull/8306

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java (hunk 11 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java]: Good point

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java [hunk 12]
```diff
@@ -1403,21 +1406,86 @@ public void shouldNotLoopInfinitelyOnMissingMetadataAndShouldNotCreateRelatedTas
...
+
+        partitionAssignor.setAssignmentErrorCode(AssignorError.NONE.code());
+        partitionAssignor.onAssignment(createAssignment(newHostState), null);
+
+        assertThat(partitionAssignor.assignmentErrorCode(), is(AssignorError.NONE.code()));
 
         EasyMock.verify(taskManager);
     }
 
+    @Test
```
Comment (abbccdda) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java]: nit: Just my usual habit, I think we should get a separate test file `HostInfoTest` instead.

**Review by abbccdda — APPROVED** [review]
The changes LGTM. With just one minor comment

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java (hunk 12 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java]: fair enough, moved to `HostInfoTest`

**Comment by guozhangwang** [comment]
test this please

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java [hunk 13]
```diff
@@ -1281,15 +1284,54 @@ public void shouldNotLoopInfinitelyOnMissingMetadataAndShouldNotCreateRelatedTas
...
     @Test
     public void shouldUpdateClusterMetadataAndHostInfoOnAssignment() {
-        final TopicPartition partitionOne = new TopicPartition("topic", 1);
-        final TopicPartition partitionTwo = new TopicPartition("topic", 2);
-        final Map<HostInfo, Set<TopicPartition>> hostState = Collections.singletonMap(
-            new HostInfo("localhost", 9090), mkSet(partitionOne, partitionTwo));
+        final Map<HostInfo, Set<TopicPartition>> initialHostState = mkMap(
+            mkEntry(new HostInfo("localhost", 9090), mkSet(t1p0, t1p1)),
+            mkEntry(new HostInfo("otherhost", 9090), mkSet(t2p0, t2p1))
+            );
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:1290]: nit: extra spaces?

**Review by guozhangwang — APPROVED** [review]
LGTM! Just nit comment. We can merge the PR as-is after green builds since I know you have more PRs touching on the test file anyways.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java (hunk 13 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:1290]: Ack, I'll fix this in a next PR

**Comment by ableegoldman** [comment]
@guozhangwang tests passed, with one unrelated failure:

`org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest.testReplication`

**Comment by guozhangwang** [comment]
Merged to trunk.
